### PR TITLE
Spec for font features and axes of variation

### PR DESCRIPTION
**Pull request by PankajBhojwani**
Add a spec for how we could allow users to define font features and axes of variation.

References #1790

### CODE REVIEW

**Review by miniksa — APPROVED** [review]
Exactly what I was hoping for!

**Comment by skyline75489** [comment]
Impressive spec! This makes me realize  DxFontInfo is more important than I think, as it also opens up the opportunity for all the other font features.

**Comment by DHowett** [comment]
nit: put the epic number in the filename!

**Comment by DHowett** [comment]
and link it in the body (in the yaml front matter)

File: doc/specs/Font features and axes-spec.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+There is also a standard list of axes of variation, and each axis has its own default. We will approach this the same way we approached font features, by allowing users to specify additional features or omit features without needing to redefine the defaults.
+
+## UI/UX Design
+
+Users will be able to add a new setting to their font objects (added in [#10433](https://github.com/microsoft/terminal/pull/10433)). The resultant font object may look something like this
+
+```
+"font": {
+    "face": "Cascadia Code",
+    "size": 12,
```
Comment (DHowett) [inline doc/specs/Font features and axes-spec.md:49]: nit: we should include weight _here_, and not force people to set "axes: wght: 350". However: we need the spec to clarify what happens when there are multiple values and they do not agree.


What is the order in which these are processed:?

`fontWeight` `font.weight` `font.axes[wght]`?

File: doc/specs/Font features and axes-spec.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
+We will need to monitor any potential performance issues caused by this feature. Since the extra work being done is only during initialization, the performance issues will probably be negligible if at all, but it will still be wise to check.
+
+## Future considerations
+
+DWrite additionally offers the ability to vary the font features across runs of text. However, for our initial implementation of this feature, we will only apply font features to the entire buffer. If/when we decide to allow specifying font features for particular runs of text, we can lean into our existing mechanisms of splitting up runs of text to implement that.
```
Comment (DHowett) [inline doc/specs/Font features and axes-spec.md:86]: This offers us some caching opportunities, no?

**Review by DHowett — CHANGES_REQUESTED** [review]

File: doc/specs/Font features and axes-spec.md (hunk 1 above)
Comment (DHowett) [reply doc/specs/Font features and axes-spec.md:49]: Oh dang, this is actually part of the other spec isn't it

File: doc/specs/Font features and axes-spec.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+### Reliability
+
+Aside from additional parsing required for the settings file (which inherently offers more locations for parsing to fail), we need to be careful about badly formed/non-existant feature tags or axes specified in the user-defined dictionaries. We must make sure to ignore such declarations (perhaps alongside emitting a warning to the user) and only apply those that are correctly formed and exist.
+
+### Compatibility
+
+Older versions of Windows may not have the DWrite updates that allow for defining font features and axes of variation. We must make sure to fallback to the current implementation in these cases.
+
+### Performance, Power, and Efficiency
+
```
Comment (carlos-zamora) [inline doc/specs/Font features and axes-spec.md:79]:
```suggestion
### Performance, Power, and Efficiency

Should not affect performance, power, or efficiency.

```
presumably, this is what you meant to do

File: doc/specs/Font features and axes-spec.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+### **Axes of variation**
+
+Specifying axes of variation is done in an extremely similar manner to the way font features are specified - a 4-character tag is used to specify which font axis is being modified and a numerical value is provided to specify the value the axis should be set to. For example, {'slnt', 20} specifies that the 'slant' axis should be set to 20.
+
+There is also a standard list of axes of variation, and each axis has its own default. We will approach this the same way we approached font features, by allowing users to specify additional features or omit features without needing to redefine the defaults.
+
+## UI/UX Design
```
Comment (carlos-zamora) [inline doc/specs/Font features and axes-spec.md:42]: Are we adding this to the Settings UI?
- yes: maybe give a rough idea of what that would look like. I think a design would be overkill, but maybe mentioning little things like where the setting will be and how we would handle all of the possible feature tags. I guess we would have to manually translate the feature tags to a more user-friendly format?
- no: just mention this this in future considerations

**Review by carlos-zamora — APPROVED** [review]
minor things

File: doc/specs/Font features and axes-spec.md (hunk 3 above)
Comment (PankajBhojwani) [reply doc/specs/Font features and axes-spec.md:79]: Thanks for catching that!

File: doc/specs/Font features and axes-spec.md (hunk 4 above)
Comment (PankajBhojwani) [reply doc/specs/Font features and axes-spec.md:42]: Mentioned it in future considerations for now!

File: doc/specs/Font features and axes-spec.md (hunk 1 above)
Comment (PankajBhojwani) [reply doc/specs/Font features and axes-spec.md]: Between `font.weight` and `fontWeight`, we will [only process `font.weight` and ignore `fontWeight`](https://github.com/microsoft/terminal/pull/10433/files#diff-bbb0e3dd566eb39683a77f3e0fa86afa4a544f72ab7a81ed732f0d48018d6ed3R58-R77). 

Between `font.weight` and `font.axes[wght]`... I'm inclined to say we should ignore `font.weight` and use `font.axes[wght]`, thoughts?

File: doc/specs/Font features and axes-spec.md (hunk 2 above)
Comment (PankajBhojwani) [reply doc/specs/Font features and axes-spec.md]: Not entirely sure what you mean here, what would we cache?

File: doc/specs/Font features and axes-spec.md (hunk 1 above)
Comment (zadjii-msft) [reply doc/specs/Font features and axes-spec.md]: I'd agree with @PankajBhojwani here. My reasoning is that if someone went through the trouble of defining `font.axes[wght]`, then _they know what they are doing_. I don't, so I'm probably gonna stick with `fontWeight`, but hey, that's the whole point of layering.

Stick this example in "Potential Issues", and call it answered

File: doc/specs/Font features and axes-spec.md (hunk 2 above)
Comment (zadjii-msft) [reply doc/specs/Font features and axes-spec.md]: yea I don't know what this is referring to

File: doc/specs/#1790 - Font features and axes-spec.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+Reference: [#1790](https://github.com/microsoft/terminal/issues/1790)
+
+Currently, if a font has ligatures, we offer no way for a user to disable them. Many users would like the option to do so, and would also like the ability to choose stylistic sets for fonts - for example, at the time of this writing, Cascadia Code offers 4 stylistic sets but we offer no way for users to specify any of them.
+
+In a similar vein, many fonts allow for setting variations on the font along certain attributes, commonly referred to as 'axes of variation'. We can offer users more font customization options by allowing them to configure these font variations.
+
+## Solution Design
+
+### **Font features**
```
Comment (zadjii-msft) [inline doc/specs/#1790 - Font features and axes-spec.md:24]: nit:

```suggestion
### Font features
```

I don't think bolding a header does anything (here and below)

File: doc/specs/#1790 - Font features and axes-spec.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+Specifying axes of variation is done in an extremely similar manner to the way font features are specified - a 4-character tag is used to specify which font axis is being modified and a numerical value is provided to specify the value the axis should be set to. For example, {'slnt', 20} specifies that the 'slant' axis should be set to 20.
+
+There is also a standard list of axes of variation, and each axis has its own default. We will approach this the same way we approached font features, by allowing users to specify additional features or omit features without needing to redefine the defaults.
+
+## UI/UX Design
+
+Users will be able to add a new setting to their font objects (added in [#10433](https://github.com/microsoft/terminal/pull/10433)). The resultant font object may look something like this
+
+```
```
Comment (zadjii-msft) [inline doc/specs/#1790 - Font features and axes-spec.md:47]: nit:
```suggestion
```json
```

to get syntax highlighting

**Review by zadjii-msft — APPROVED** [review]
Yea I dig this spec. Sounds like you'll be able to close out almost all of #1790 in one fell swoop then. Add the comment about the order of `fontWeight`, `font.weight`, and `font.axes[wght]`, and this is good2go

File: doc/specs/Font features and axes-spec.md (hunk 2 above)
Comment (DHowett) [reply doc/specs/Font features and axes-spec.md]: I DON'T EITHER OKAY 😁

File: doc/specs/#1790 - Font features and axes-spec.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+Older versions of Windows may not have the DWrite updates that allow for defining font features and axes of variation. We must make sure to fallback to the current implementation in these cases.
+
+### Performance, Power, and Efficiency
+
+Should not affect power or efficiency. See the Potential Issues section for potential performance issues.
+
+## Potential Issues
+
+We will need to monitor any potential performance issues caused by this feature. Since the extra work being done is only during initialization, the performance issues will probably be negligible if at all, but it will still be wise to check.
+
```
Comment (DHowett) [inline doc/specs/#1790 - Font features and axes-spec.md:86]: I'd rather you address the performance issue in the performance section! We discovered that we could *no longer* use "IsSimple" to figure out a text run, so we had to run everything through the analyzer if there were non-default font features. That is a **BIG DEAL** that we will need to make sure is called out in this spec!

**Review by DHowett — CHANGES_REQUESTED** [review]

File: doc/specs/#1790 - Font features and axes-spec.md (hunk 7 above)
Comment (PankajBhojwani) [reply doc/specs/#1790 - Font features and axes-spec.md]: Good call, added!

**Review by DHowett — APPROVED** [review]

**Pull request merged**